Zend/zend_gc.c

```diff
@@ -245,6 +245,56 @@ static zend_gc_globals gc_globals;
 # define GC_TRACE(str)
 #endif
 
+
+#define GC_STACK_SEGMENT_SIZE (((4096 - ZEND_MM_OVERHEAD) / sizeof(void*)) - 2)
+
+typedef struct _gc_stack gc_stack;
+
+struct _gc_stack {
+	gc_stack        *prev;
+	gc_stack        *next;
+	zend_refcounted *data[GC_STACK_SEGMENT_SIZE];
+};
+
+#define GC_STACK_DCL(init) \
+	gc_stack *_stack = init; \
+	size_t    _top = 0;
+
+#define GC_STACK_PUSH(ref) \
+	do { \
+		if (UNEXPECTED(_top == GC_STACK_SEGMENT_SIZE)) { \
+			if (UNEXPECTED(!_stack->next)) { \
+				gc_stack *segment = emalloc(sizeof(gc_stack)); \
```

Comment:
If there is a need for bigger and bigger stack, can we double the segment size with each allocation?
I think that this way the number of allocations will increase logarithmically with the number of objects in stack.

Reply:
In general,  we can, but allocation of 4096 page is simpler for Zend MM.

On the same benchmark, logarithmic erealloc() was called 108 times, and emallc/efree 1080 times, but taking slightly less CPU instructions, anyway.

In any case the impact of emalloc/efree on zend_gc_collect_cycles() is less than 0.2%.

Reply:
ok, got it.